api/v1alpha4/azurecluster_default.go

```diff
@@ -41,91 +41,91 @@ const (
 )
 
-func (c *AzureCluster) setDefaults() {
-	c.setResourceGroupDefault()
-	c.setAzureEnvironmentDefault()
-	c.setNetworkSpecDefaults()
+func (s *AzureClusterSpec) setDefaults(name string) {
+	s.setResourceGroupDefault(name)
+	s.setAzureEnvironmentDefault()
+	s.setNetworkSpecDefaults(name)
 }
 
-func (c *AzureCluster) setNetworkSpecDefaults() {
-	c.setVnetDefaults()
-	c.setBastionDefaults()
-	c.setSubnetDefaults()
-	c.setAPIServerLBDefaults()
-	c.setNodeOutboundLBDefaults()
-	c.setControlPlaneOutboundLBDefaults()
+func (s *AzureClusterSpec) setNetworkSpecDefaults(name string) {
+	s.setVnetDefaults(name)
+	s.setBastionDefaults(name)
+	s.setSubnetDefaults(name)
+	s.setAPIServerLBDefaults(name)
+	s.setNodeOutboundLBDefaults(name)
+	s.setControlPlaneOutboundLBDefaults(name)
 }
 
-func (c *AzureCluster) setResourceGroupDefault() {
-	if c.Spec.ResourceGroup == "" {
-		c.Spec.ResourceGroup = c.Name
+func (s *AzureClusterSpec) setResourceGroupDefault(name string) {
+	if s.ResourceGroup == "" {
+		s.ResourceGroup = name
 	}
 }
 
-func (c *AzureCluster) setAzureEnvironmentDefault() {
-	if c.Spec.AzureEnvironment == "" {
-		c.Spec.AzureEnvironment = DefaultAzureCloud
+func (s *AzureClusterSpec) setAzureEnvironmentDefault() {
+	if s.AzureEnvironment == "" {
+		s.AzureEnvironment = DefaultAzureCloud
 	}
 }
 
-func (c *AzureCluster) setVnetDefaults() {
-	if c.Spec.NetworkSpec.Vnet.ResourceGroup == "" {
-		c.Spec.NetworkSpec.Vnet.ResourceGroup = c.Spec.ResourceGroup
+func (s *AzureClusterSpec) setVnetDefaults(name string) {
+	if s.NetworkSpec.Vnet.ResourceGroup == "" {
+		s.NetworkSpec.Vnet.ResourceGroup = s.ResourceGroup
 	}
-	if c.Spec.NetworkSpec.Vnet.Name == "" {
-		c.Spec.NetworkSpec.Vnet.Name = generateVnetName(c.ObjectMeta.Name)
+	if s.NetworkSpec.Vnet.Name == "" {
+		s.NetworkSpec.Vnet.Name = generateVnetName(name)
 	}
-	if len(c.Spec.NetworkSpec.Vnet.CIDRBlocks) == 0 {
-		c.Spec.NetworkSpec.Vnet.CIDRBlocks = []string{DefaultVnetCIDR}
+	if len(s.NetworkSpec.Vnet.CIDRBlocks) == 0 {
+		s.NetworkSpec.Vnet.CIDRBlocks = []string{DefaultVnetCIDR}
 	}
 }
 
-func (c *AzureCluster) setSubnetDefaults() {
-	cpSubnet, err := c.Spec.NetworkSpec.GetControlPlaneSubnet()
+func (s *AzureClusterSpec) setSubnetDefaults(name string) {
+	cpSubnet, err := s.NetworkSpec.GetControlPlaneSubnet()
 	if err != nil {
 		cpSubnet = SubnetSpec{Role: SubnetControlPlane}
-		c.Spec.NetworkSpec.Subnets = append(c.Spec.NetworkSpec.Subnets, cpSubnet)
+		s.NetworkSpec.Subnets = append(s.NetworkSpec.Subnets, cpSubnet)
 	}
 
-	nodeSubnet, err := c.Spec.NetworkSpec.GetNodeSubnet()
+	nodeSubnet, err := s.NetworkSpec.GetNodeSubnet()
 	if err != nil {
 		nodeSubnet = SubnetSpec{Role: SubnetNode}
-		c.Spec.NetworkSpec.Subnets = append(c.Spec.NetworkSpec.Subnets, nodeSubnet)
+		s.NetworkSpec.Subnets = append(s.NetworkSpec.Subnets, nodeSubnet)
 	}
 
 	if cpSubnet.Name == "" {
-		cpSubnet.Name = generateControlPlaneSubnetName(c.ObjectMeta.Name)
+		cpSubnet.Name = generateControlPlaneSubnetName(name)
```

Comment:
I made code changes so that the template and the cluster types have the same defaulting and validation rules. However I am not sure if the template type should have all the defaulting rules (and validations) that the cluster has. 
In this case, when we use `setSubnetDefaults` on the template type we use the template name as the argument and therefore the subnetname is based on the clustertemplate name. When clusters are created using this template they will all share the same subnet name. Is that okay?

Reply:
Good point. As long as they are in separate resource groups, they should be fine. But, the resource group also relies on the cluster template name. So, we might have to find a way around for that.

Reply:
Should we even set defaults for the template? the generated name for azure cluster will be different from the template name.

Reply:
> Should we even set name related defaults for the template?

I have similar doubts. As long as all the validations/defaulting we perform on the template are also performed on the cluster we wont end up in the continuous reconcile loop (i think). 
The validation/defaulting we have on template should be a subset of the cluster type. But the cluster might have more validation/defaulting on top of what the template already does. 
As long as the validation/defaulting do not conflict with the ones from the template we are probably good.

cc @fabriziopandini thoughts?

Reply:
One way to work around this can be to add an annotation in azure cluster to signal that it was created from a template and handle defaults accordingly.

Reply:
The ClusterClass proposal already suggests adding the following label
```yaml
cluster.x-k8s.io/topology: <cluster-name>
```
to the infrastructure cluster object. We can use this to detect if the cluster was created from a template.

Note that this only works for the current usecases (ClusterClass realm) we have and does not work for any future use cases outside of the clsuterclass workflow where infrastructure cluster objects would be created out of the template object.

Even if we do detect if the cluster is created using a template we might still need to have all the defaulting and validation that a normal cluster object would have. I believe it is the defaulting and validations of the template that might need to be adjusted.

Reply:
I'm not sure I have all the context here, but I'm happy to talk about this; In the meantime let me refocus on the problem statement and some of the above comments.

ClusterClass requirements and expected behaviour:

1 For creating a CAPZ cluster using ClusterClass, it is necessary to introduce the AzureClusterTemplate.
2 The spec.template from AzureClusterTemplate will be used to create actual AzureCluster, and we should make sure this operation can succeed.
3 After creating the actual AzureCluster, ClusterClass will continuously reconcile its content to the original spec.template from AzureClusterTemplate, and we should make sure to avoid continuous reconciliation or other unexpected behaviour, like value override.
4 ClusterClass, who will manage 2 and 3, will not have any knowledge of the Azure specific types, and treat them as unstructured

> The validation/defaulting we have on template should be a subset of the cluster type. But the cluster might have more validation/defaulting on top of what the template already does.

I'm not sure about this. 
more validation rules in the cluster than in the template: what if someone sets a value in the template (no validation rule) that won't be accepted in the cluster (additional validation rule)? this will prevent 2 to successfully create a Cluster.

more defaulting rules in the cluster than in the template:  if you leave a field empty in the template (no defaulting rule), and the you default it in the actual cluster (additional validation rule), then 3 will continuously reconcile and try to set this value  back to empty.

> One way to work around this can be to add an annotation in azure cluster to signal that it was created from a template and handle defaults accordingly

This could make sense to me, but it really depends by the specific problem
There are also other options we can consider, like making 4 more smart, but this has to be discussed and the solution should be generic enough to work will all the providers

Reply:
need to spend more time thinking about this but my initial thought on this is that we should _not_. share RG/vnet name/subnet names between AzureClusters created from the same template.

Reply:
cc: @randomvariable @yastij as there could be similar concerns in other providers.

Reply:
This issue will be addressed according to the recommendations listed here: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/202105256-cluster-class-and-managed-topologies.md#conventions-for-template-types-implementation

TLDR; `AzureClusterTemplate` will only contain a subset of fields of the `AzureCluster`. Other fields in `AzureCluster` that need to be set (that are cluster specific) will be set using defaulting. 
Note: This will not trigger a continuous reconciliation. Refer to:
>If a field does not have a value in the template, but the users/another controller sets this value in the object, the reconcile topology process will preserve the field value.


cc @shysank @CecileRobertMichon